<a href="https://colab.research.google.com/github/Angelysa5/Anal-tica/blob/main/Reto_1_Modelos_de_Clasificaci%C3%B3n_(Naive_Bayes)_Angely_Salazar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1 – Modelos de Clasificación (Naive Bayes)**

De acuerdo con la base de datos Longitudinal Survey (Apendice A, Conjunto 5: Datos de Levantamiento Longitudinal del Libro Guía Analítica de Negocios Comunicación con Datos), llevar a cabo la implementación de un modelo Naive Bayes para determinar si una persona vive en un área urbana o en un área rural.


In [ ]:
#Se carga el archivo de excel con la base de datos a trabajar.
#Se procede con la carga de las librerías de trabajo.

import numpy as np #Librería que permite las funciones matemáticas
import pandas as pd #Librería que permite comunicación con Excel
from sklearn.naive_bayes import GaussianNB #función Gaussiana
from sklearn.metrics import confusion_matrix #La métrica de medida

1. Se procede con la carga de los datos.

In [ ]:
nxl = '/content/1. BD1_Longitudinal Survey_Int.xlsx'

XDB=pd.read_excel(nxl,sheet_name=0) #Se traen los datos necesarios = DataFrame

#Seleccionamos las variables de interés =
#Variables socioeconómicas como: Edad, Educación de los Padres, Estatura, Peso, Ingresos, Escala de Autoestima, Tamaño de la Familia.

XDB=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]

XDB=XDB.dropna() #Eliminar datos faltantes, celdas vacías
print(XDB)


       Age  Mother_Edu  Father_Edu  Height  Weight   Income  Self_Esteem  \
1       20         5.0         8.0    62.0   120.0      0.0         16.0   
5       19        12.0        12.0    64.0   200.0  40000.0         27.0   
6       15        12.0        12.0    65.0   131.0  25000.0         26.0   
7       21         9.0         6.0    65.0   179.0  27400.0         23.0   
8       16        12.0        10.0    66.0   145.0  52000.0         26.0   
...    ...         ...         ...     ...     ...      ...          ...   
12557   20        12.0        12.0    60.0   110.0   1000.0         20.0   
12565   20        12.0        11.0    63.0   152.0  38000.0         24.0   
12647   21        16.0        12.0    70.0   180.0  43600.0         30.0   
12651   21        16.0        16.0    73.0   170.0  40000.0         27.0   
12678   21         8.0         8.0    70.0   180.0  17000.0         19.0   

       FamilySize  Urban  
1               5    1.0  
5               4    1.0  
6     

In [ ]:
total_datos = len(XDB)
print("En la base de datos hay", total_datos, "datos.")

En la base de datos hay 5948 datos.


**¿Cuál es el número de datos total de la base de datos?**

R= Como lo muestra por pantalla el código, hay 5.948 datos en total.

In [ ]:
#Convertir el dataframe a números
XD = np.array(XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']])

#Datos de Salida - Valor Deseado - que quiero que aprenda el modelo
yd = np.array(XDB[['Urban']])



#Porcentaje Urbano
len(XD)
len(np.where(yd==1)[0])

len(np.where(yd==1)[0])/len(XD)
print("El porcentaje de los datos de personas que viven en el área urbana es de:", len(np.where(yd==1)[0])/len(XD))


#Porcentaje Rural
len(XD)
len(np.where(yd==0)[0])

len(np.where(yd==0)[0])/len(XD)
print("El porcentaje de los datos de personas que viven en el área rural es de:", len(np.where(yd==0)[0])/len(XD))

El porcentaje de los datos de personas que viven en el área urbana es de: 0.7916946872898454
El porcentaje de los datos de personas que viven en el área rural es de: 0.20830531271015468


**¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?**

El 79.1% vive en el área URBANA.

El 20.8% vive en el área RURAL.

In [ ]:
#Se procede a implementar el modelo
mnb = GaussianNB() #Llamamos al modelo
mnb.fit(XD,yd)
ydp = mnb.predict(XD) #Muestra los datos de salida QUE ÉL OBTUVO



#Parámetros del Modelo por Categoría - Recordar = Rural 0 (arriba)- Urbano 1(abajo)

u = mnb.theta_ #Dígame las MEDIAS de las variables para cada categoría
#u = miu

sigma = np.sqrt(mnb.var_) #Dígame las VARIANZAS de las variables para cada categoría


#Variaciones
LI = u-sigma #Límite inferior
LS = u+sigma #Límite superior

print("La medida de las variables por categoría:")
print(u)
print()
print("El límite inferior de la variación de las variables de la categoría es:")
print(LI)
print()
print("El límite superior de la variación de las variables de la categoría es:")
print(LS)


La medida de las variables por categoría:
[[1.83640032e+01 1.07312349e+01 1.02711864e+01 6.72429379e+01
  1.49003228e+02 3.01248854e+04 2.19951574e+01 4.73930589e+00]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 6.69116585e+01
  1.44750903e+02 3.32736742e+04 2.25744319e+01 4.68103631e+00]]

El límite inferior de la variación de las variables de la categoría es:
[[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  6.31606515e+01
   1.18654593e+02 -1.52335198e+03  1.77786493e+01  2.18703841e+00]
 [ 1.58884167e+01  7.76073072e+00  7.11767172e+00  6.27097275e+01
   1.14722372e+02 -2.81605593e+03  1.83768011e+01  2.23897780e+00]]

El límite superior de la variación de las variables de la categoría es:
[[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.13252242e+01
  1.79351863e+02 6.17731228e+04 2.62116655e+01 7.29157338e+00]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.11135895e+01
  1.74779433e+02 6.93634044e+04 2.67720628e+01 7.12309482e+00]]


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


**Determinar el intervalo en el cual se encuentra la variación de la Edad**

R = 15-20 años

La edad de las personas que viven en el área RURAL está entre los *15 y los 20* años.

La edad de las personas que viven en el área URBANA está entre los *15 y los 20* años.

**Determinar el intervalo en el cual se encuentra la variación de la Altura**

R = Rural 63-71 y Urbano 62-71.

La altura de las personas que viven en el área RURAL está entre los *63 y 71*.

La altura  de las personas que viven en el área URBANA está entre los *62 y 71*.

**Determinar el intervalo en el cual se encuentra la variación del Peso de las Personas**

R = Rural 118-179 y Urbano 114-174

El peso de las personas que viven en el área RURAL está entre 118 y 179

El peso de las personas que viven en el área URBANA está entre 114 y 174

In [ ]:
#Métrica de Evaluación - cómo se comportó el modelo identificando si es rural o urbano

#Matriz de Confusión
cm = confusion_matrix(yd,ydp)
#CTR + SHIFT + ENTER

array([[   3, 1236],
       [   8, 4701]])

In [ ]:
#Se procede con la evaluación del modelo
VN = cm[0,0] #Verdaderos Negativos
FP = cm[0,1] #Falsos Positivos
FN = cm[1,0] #Falsos Negativos
VP = cm[1,1] #Verdaderos Positivos

Exactitud = (VP+VN)/len(XD)
print("La Exactitud es: ", Exactitud) #Comportamiento General

#Modelo que esté superior al 75% está súper bien

#Tasa de Error
TE = (FN+FP)/len(XD)
print("La Tasa de Error es: ", TE) #Porcentaje Equivocaciones que tuvo el modelo

Sensibilidad = VP/(VP+FP)
print("La Sensibilidad es:", Sensibilidad) #Cómo se comportó frente a los positivos

Especificidad = VN/(VN+FN)
print("la Especificidad es:", Especificidad) #Cómo se comportó frente a los negativos


La Exactitud es:  0.7908540685944856
La Tasa de Error es:  0.20914593140551446
La Sensibilidad es: 0.7918140474987367
la Especificidad es: 0.2727272727272727


**¿Cuál es la exactitud alcanzada por el modelo?**

R = 79%

**¿Cuál es la tasa de error alcanzada por el modelo?**

R = 20%

**¿Cuál es la Sensibilidad alcanzada por el modelo?**

R = 79%

**¿Cuál es la Especificidad alcanzada por el modelo?**

R = 27%


In [ ]:
Xun = np.array([[34,9,6,65,135,50000,30,3]]) #estas son las variables particulares de la persona del Excel que deseo analizar
ydun = mnb.predict(Xun)
print(ydun)

#Como se observa por pantalla, esta persona pertenece al grupo de 1, es decir, es una persona que vive en la CIUDAD.

[1.]


**¿Qué pasa cuando deseo analizar una sola persona?**

Para una persona con las siguientes características: ['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']=[34,9,6,65,135,50000,30,3]

¿vive en la ciudad o en el área rural?

**R =** Una persona con dichas características vive en la **CIUDAD.**

**ANÁLISIS DE RESULTADOS:**

De los resultados arrojados por el modelo que aborda 5.948 datos de los cuales el 20.8% hace referencia a aquellos usuarios que viven en un área rural y 79.1% a aquellos que viven en un área urbana, se puede observar que la variación de la edad a analizar se encuentra entre los 15 y los 20 años, pero en términos generales, aunque no es una diferencia demasiado abismal, las personas que viven en las áreas rurales respecto a la altura(63-71) y el peso(118-179), tiende a tener valores más altos que aquellas personas que viven en el área urbana las cuales tienen una altura entre (62-71) y un peso entre (114-174).

Con respecto a las métricas se puede observar que el modelo funcionó mejor clasificando los positivos, como lo dice el indicador de Sensibilidad el cual alcanzó el 79% de certeza.

Analizando en conjunto, se puede observar que el modelo logró una exactitud del 79% lo cual es un excelente indicador ya que supera el mínimo 72% esperado para poder ser determinado como un buen porcentaje de exactitud que a su vez lo convierte en un buen modelo, además, se obtuvo una especificidad del 27% y una tasa de error cercana al 20% en promedio.

Con respecto a la persona de la cual no se tiene certeza de si vive en la ciudad o en el campo, al caracterizarse por tener una edad de 34 años, una madre con un nivel educativo de 9, un padre con un nivel educativo de 6, una altura de 65, un peso de 135, unos ingresos de 50.000, una escala de autoestima de 30 y un tamaño de familia de 3 miembros, se logra concluir que este personaje pertenece a la categoría 1, lo que nos indica que vive en el área urbana.